### PR TITLE
Snippets pkg setters

### PR DESCRIPTION
**Pull request by yuzhouliu9**
Added constructor and setters.
Replaced the old complicated manual Type construction with testdata directory.

### CODE REVIEW

File: pkg/generators/snippets/constructor.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package snippets
+
+import (
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/types"
+)
+
+func GenerateConstructor(t *types.Type) (string, generator.Args) {
```
Comment (dskatz) [inline pkg/generators/snippets/constructor.go:8]: The name should specify generating a constructor with ObjectMeta

File: pkg/generators/snippets/setters.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
+package snippets
+
+import (
+	"fmt"
+
+	"k8s.io/gengo/generator"
+	"k8s.io/gengo/types"
+)
+
+func GenerateSetterPrimitive(root *types.Type, member types.Member) (string, generator.Args) {
```
Comment (dskatz) [inline pkg/generators/snippets/setters.go:10]: nit: GenerateSetterForPrimitiveType

File: pkg/generators/snippets/setters.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+func (o *$.type|raw$) $.funcName$(in $.memberType|raw$) *$.type|raw$ {
+	o.$.memberName$ = in
+	return o
+}
+
+`
+	return raw, args
+}
+
+func GenerateSetterMap(root *types.Type, member types.Member) (string, generator.Args) {
```
Comment (dskatz) [inline pkg/generators/snippets/setters.go:28]: nit : GenerateSetterForMap

File: pkg/generators/snippets/setters.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+func (o *$.type|raw$) $.funcName$(in $.memberType|raw$) *$.type|raw$ {
+	o.$.memberName$ = mergeMapStringString(o.$.memberName$, in)
+	return o
+}
+
+`
+	return raw, args
+}
+
+func GenerateSetterSlice(root *types.Type, member types.Member, wrapper *types.Type) (string, generator.Args) {
```
Comment (dskatz) [inline pkg/generators/snippets/setters.go:46]: nit: GenerateSetterForSlice

File: pkg/generators/snippets/setters.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+	args := generator.Args{
+		"type":       root,
+		"funcName":   funcName(member),
+		"memberName": member.Name,
+	}
+
+	var raw string
+	if wrapper != nil {
+		// needs translation from struct to pull out the embedded type
```
[the remark this answers is not in the document]
Comment (yuzhouliu9) [reply pkg/generators/snippets/setters.go:55]: An example of this is clusterrole.go [AppendRules](https://github.com/10gen/kanopy/blob/master/pkg/builder/k8s/clusterrole.go#L30)

`PolicyRule` embeds `rbacv1.PolicyRule`. In order to append the input to the slice of `rbacv1.PolicyRule`, the code has to loop over the input and translate/extract the `PolicyRule`.

For this example, the input arguments to this function would be:
`GenerateSetterForSlice(root=ClusterRole, member=Rules, wrapper=PolicyRule)`

I hope that makes sense. I can change the comment to make it more clear

File: pkg/generators/snippets/setters.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+func GenerateSetterStruct(root *types.Type, member types.Member, wrapper *types.Type) (string, generator.Args) {
+	args := generator.Args{
+		"type":       root,
+		"funcName":   funcName(member),
+		"memberName": member.Name,
+	}
+
+	var raw string
+	if wrapper != nil {
+		// needs translation from struct to pull out the embedded type
```
[the remark this answers is not in the document]
Comment (yuzhouliu9) [reply pkg/generators/snippets/setters.go:91]: Similar to answer before.
A example of this where input argument is actually a wrapper struct that embeds a k8s.io struct is https://github.com/10gen/kanopy/blob/master/pkg/builder/k8s/pod.go#L79

File: pkg/generators/snippets/setters.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (dskatz) [reply pkg/generators/snippets/setters.go:55]: That does.   I think then that the methods that contain a `wrapper` argument need to split into two functions.

1. `GenerateSetterForEmbeddedSlice`   - Can be used to cover a `type MyStruct { a.AnotherStruct }`
2. `GenerateSetterForMemberSlice`.      - Used to cover a struct `type MemberStruct { Name string}`

File: pkg/generators/snippets/setters_test.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+	tests := []struct {
+		ctx     *generator.Context
+		root    *types.Type
+		member  types.Member
+		wrapper *types.Type
+		want    string
+	}{
+		{
+			// embedded type
```
[the remark this answers is not in the document]
Comment (yuzhouliu9) [reply pkg/generators/snippets/setters_test.go]: thanks. After splitting to `GenerateSetterForEmbeddedSlice` and `GenerateSetterForMemberSlice `, table tests are not necessary anymore, so I removed the comments/description as well.

File: pkg/generators/snippets/setters_test.go [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+			ctx:     ctx,
+			root:    someStruct,
+			member:  someStruct.Members[0].Type.Members[a.MemberIndex_SomeStruct_CStruct], // SomeStruct.SomeStruct.CStruct
```
[the remark this answers is not in the document]
Comment (yuzhouliu9) [reply pkg/generators/snippets/setters_test.go]: added helper `getMemberFromType` which takes in a type, and a variadic list of selector strings. Will drill down to the member using the list of selector strings.

File: pkg/generators/snippets/setters.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yuzhouliu9) [reply pkg/generators/snippets/setters.go]: Ok thanks, I split them up

**Review by dskatz — APPROVED** [review]
LGTM.

File: pkg/generators/snippets/helper_test.go [hunk 9]
```diff
@@ -40,44 +45,47 @@ func newTestGeneratorContext() (*generator.Context, error) {
...
+	d.InputDirs = []string{dir + "/..."}
+
+	b, err := d.NewBuilder()
+	assert.NoError(t, err)
+
+	findTypes, err := b.FindTypes()
+	assert.NoError(t, err)
+
+	n := findTypes[dir].Types[selector]
+	fmt.Println(findTypes, n.Members)
```
[the remark this answers is not in the document]
Comment (yuzhouliu9) [reply pkg/generators/snippets/helper_test.go:61]: woops great catch, thanks

**Pull request merged**